library/src/commonMain/kotlin/com/badoo/reaktive/completable/DoOnBefore.kt

```diff
@@ -0,0 +1,109 @@
+package com.badoo.reaktive.completable
+
+import com.badoo.reaktive.disposable.Disposable
+import com.badoo.reaktive.disposable.wrap
+import com.badoo.reaktive.utils.lock.newLock
+import com.badoo.reaktive.utils.lock.synchronized
+
+fun Completable.doOnBeforeSubscribe(action: (Disposable) -> Unit): Completable =
+    completable { observer ->
+        subscribeSafe(
+            object : CompletableObserver by observer {
+                override fun onSubscribe(disposable: Disposable) {
+                    action(disposable)
+                    observer.onSubscribe(disposable)
+                }
+            }
+        )
+    }
+
+fun Completable.doOnBeforeComplete(action: () -> Unit): Completable =
+    completable { observer ->
+        subscribeSafe(
+            object : CompletableObserver by observer {
+                override fun onComplete() {
+                    action()
+                    observer.onComplete()
+                }
+            }
+        )
+    }
+
+fun Completable.doOnBeforeError(consumer: (Throwable) -> Unit): Completable =
+    completable { observer ->
+        subscribeSafe(
+            object : CompletableObserver by observer {
+                override fun onError(error: Throwable) {
+                    consumer(error)
+                    observer.onError(error)
+                }
+            }
+        )
+    }
+
+fun Completable.doOnBeforeTerminate(action: () -> Unit): Completable =
+    completable { observer ->
+        subscribeSafe(
+            object : CompletableObserver by observer {
+                override fun onComplete() {
+                    action()
+                    observer.onComplete()
+                }
+
+                override fun onError(error: Throwable) {
+                    action()
+                    observer.onError(error)
+                }
+            }
+        )
+    }
+
+fun Completable.doOnBeforeDispose(action: () -> Unit): Completable =
+    completable { observer ->
+        subscribeSafe(
+            object : CompletableObserver by observer {
+                override fun onSubscribe(disposable: Disposable) {
+                    observer.onSubscribe(disposable.wrap(onBeforeDispose = action))
+                }
+            }
+        )
+    }
+
+fun Completable.doOnBeforeFinally(action: () -> Unit): Completable =
```

Comment:
Could you me the difference between this one and beforeTerminate? It sort of seems like they should a single method.

Reply:
Same with RxJava, terminate is when stream reached the end (onComplete, onError), finally is when stream finished (either reached the end or disposed)